File: 3683-find-the-lexicographically-largest-string-from-the-box-i/3683-find-the-lexicographically-largest-string-from-the-box-i.py
class Solution:
    def answerString(self, word: str, numFriends: int) -> str:
        n = len(word)
        if numFriends == 1:
        # Only one friend, the entire word is collected
            return word
        elif numFriends == n:
        # Each friend gets one character; return the lex largest character
            return max(word)
        else:
            L = n - numFriends + 1  # Maximum possible length of a part
        # Find the lex largest character in the word
            max_char = max(word)
            max_substring = ''
        # Positions where the character is equal to the max character
            positions = [i for i, c in enumerate(word) if c == max_char]
            for i in positions:
            # Maximum possible end index for the substring
                end = min(i + L, n)
            # Extract the substring starting at position i with length up to L
                s = word[i:end]
                if s > max_substring:
                    max_substring = s
            return max_substring

File: 3683-find-the-lexicographically-largest-string-from-the-box-i/3683-find-the-lexicographically-largest-string-from-the-box-i.cpp
class Solution {
public:
    string answerString(string word, int numFriends) {
        int n = word.length();
        if (numFriends == 1) return word;
        if (numFriends == n) {
            char max_char = *max_element(word.begin(), word.end());
            return string(1, max_char);
        }
        int L = n - numFriends + 1;
        char max_char = *max_element(word.begin(), word.end());
        string max_substring = "";
        vector<int> positions;
        for (int i = 0; i < n; ++i) if (word[i] == max_char) positions.push_back(i);
        for (int i : positions) {
            int end = min(i + L, n);
            string s = word.substr(i, end - i);
            if (s > max_substring) max_substring = s;
        }
        return max_substring;
    }
};